### ISSUE TITLE
nn.Batchnormalization parameters

### ISSUE THREAD

**Issue by arthitag** [torch/nn#1256]
Hi,
I have a trained nngraph model(model0). I created a new nngraph model(model1) from file. Model1 has almost similar structure as model0(few extra CaddTable() layers). I copied the weights and biases of model0 to model1 using :parameters() and copy() sequentially. Model0 and model1 are giving different outputs in forward pass for the same input.

I checked the weights and biases and outputs of intermediate layer and discovered that 
for the two models, output of an intermediate Convolution layer is the same. However the output of the following BatchNormalization(BN) layer is different despite the weight and bias of the BN layer being the same(which I copied from model0 to model1). Could the problem be due to running mean and running variance? Why are the BN layers giving different outputs?

Thank you for any help.

**Comment by Naruto-Sasuke** [comment]
I am wondering about it these days too. As it can been seem from the inside code, when you save the net into your disk, when often calls `net:clearState()`, then the BN call it. Some info will be discarded like `save_mean` and `save_std` etc. However the BN layer still contains something useful like `weight`, `bias` , `running_mean` and `running_var`. 
In the in the forward of BN, it goes like this:
```lua
function BN:updateOutput(input)
   self:checkInputDim(input)

   input = makeContiguous(self, input)
   input = makeBatch(self, input)

   self.save_mean = self.save_mean or input.new()
   self.save_mean:resizeAs(self.running_mean)
   self.save_std = self.save_std or input.new()
   self.save_std:resizeAs(self.running_var)

   input.THNN.BatchNormalization_updateOutput(
      input:cdata(),
      self.output:cdata(),
      THNN.optionalTensor(self.weight),
      THNN.optionalTensor(self.bias),
      self.running_mean:cdata(),
      self.running_var:cdata(),
      self.save_mean:cdata(),
      self.save_std:cdata(),
      self.train,
      self.momentum,
      self.eps)
   return self.output
end
```
and in backward:
```lua
local function backward(self, input, gradOutput, scale, gradInput, gradWeight, gradBias)
   self:checkInputDim(input)
   self:checkInputDim(gradOutput)
   assert(self.save_mean and self.save_std, 'must call :updateOutput() first')

   input, gradOutput = makeContiguous(self, input, gradOutput)
   input = makeBatch(self, input)
   gradOutput = makeBatch(self, gradOutput)

   scale = scale or 1
   if gradInput then
      gradInput:resizeAs(gradOutput)
   end

   input.THNN.BatchNormalization_backward(
      input:cdata(),
      gradOutput:cdata(),
      THNN.optionalTensor(gradInput),
      THNN.optionalTensor(gradWeight),
      THNN.optionalTensor(gradBias),
      THNN.optionalTensor(self.weight),
      self.running_mean:cdata(),
      self.running_var:cdata(),
      self.save_mean:cdata(),
      self.save_std:cdata(),
      self.train,
      scale,
      self.eps)

   return self.gradInput
end
```
So `running_var`, `running_mean`,`save_mean`, `save_std` seems all play their roles in forward and backward. However, BN clearState:
```lua
function BN:clearState()
   -- first 5 buffers are not present in the current implementation,
   -- but we keep them for cleaning old saved models
   nn.utils.clear(self, {
      'buffer',
      'buffer2',
      'centered',
      'std',
      'normalized',
      '_input',
      '_gradOutput',
      'save_mean',
      'save_std',
   })
   return parent.clearState(self)
end
```
`save_mean` and `save_std` have been discarded. It is strange. In the other hand, in BN docs, 
```md
During training, this layer keeps a running estimate of its computed mean and std. 
The running sum is kept with a default momentum of 0.1 (unless over-ridden) 
During evaluation, this running mean/std is used for normalization.
```
So you can just try cloning `weight`,`bias`,`running_var` and `running_mean`.
Let me know whether it work or not!

**Comment by arthitag** [comment]
Hi @Naruto-Sasuke , Thanks for the reply.
I did copy all 4 parameters i.e. weight, bias, running_mean and running_std. But still the outputs of the BN layer turned out to be different for the two models. I also checked them for being same afterwards and they are! The variance(running_std) was nil for the trained model so for the new model it’s nil too. All the other parameters and input to the layer are the same.

**Comment by Naruto-Sasuke** [comment]
There is no `running_std` just `running_var`. Mention the differences of `save_std` and `running_var`

**Comment by arthitag** [comment]
Hi, thanks for pointing that out. I think this has solved the problem, at least partially. There are more recursive sub-structures in the network which has Batchnorm layer descendants, for which I need to copy the mean and variance too. If I access the nodes using model.forwardnodes, I am not able to reach these nested BN layers and forwardnodes is only returning the containing nn.Sequential() layer. Do you know a way for me to access the BN layers nested under the nn.Seuential()? I can see them when I print them on the terminal and using nnquery descendants() I can access these BN layers. But I am still struggling to get hold of their running_mean and running_var data. 

Thanks a lot for the help so far! I really appreciate it. :)

**Comment by arthitag** [comment]
Hi,
I figured out the way to recursively descend to the layers within nn.Sequential() using node.data.module.modules. Copying all the running_mean and running_var has brought the output of the network within the correct range now, although it doesn't exactly match the other(trained) network yet. But the numbers are close. 
In most of the discussions I have seen so far on copying model weights to another model, no one mentions this caveat about the extra parameters(mean and var) in Batchnorm layers that can totally mess up the outputs if not copied correctly.

**Comment by Naruto-Sasuke** [comment]
I think it it better to leave this issue open. Hope this will help more people in copying weights between models with BN layers.
